core/shared/src/main/scala/zio/FiberRefs.scala

```diff
@@ -109,78 +131,80 @@ final class FiberRefs private (
-              else
-                parentFiberRefs.updated(
-                  ref,
-                  Value(::(StackEntry(fiberId, ref.join(ref.initial, childValue), 0), List.empty), 1)
-                )
-            } { case Value(parentStack, parentDepth) =>
-              @tailrec
-              def findAncestor(
-                parentStack: List[StackEntry[?]],
-                parenthDepth: Int,
-                childStack: List[StackEntry[?]],
-                childDepth: Int
-              ): Any =
-                (parentStack, childStack) match {
-                  case (
-                        StackEntry(parentFiberId, parentValue, parentVersion) :: parentAncestors,
-                        StackEntry(childFiberId, childValue, childVersion) :: childAncestors
-                      ) =>
-                    if (parentFiberId == childFiberId)
-                      if (childVersion > parentVersion) parentValue else childValue
-                    else if (childDepth > parentDepth)
-                      findAncestor(parentStack, parentDepth, childAncestors, childDepth - 1)
-                    else if (childDepth < parentDepth)
-                      findAncestor(parentAncestors, parentDepth - 1, childStack, childDepth)
-                    else
-                      findAncestor(parentAncestors, parentDepth - 1, childAncestors, childDepth - 1)
-                  case _ =>
-                    (ref.initial)
-                }
-              val childValue = childHead.value
-
-              val ancestor = findAncestor(parentStack, parentDepth, childStack, childDepth)
-
-              val patch = ref.diff(ancestor, childValue)
-
-              val oldValue = parentStack.head.value
-              val newValue = ref.join(oldValue, ref.patch(patch)(oldValue))
-
-              if (oldValue == newValue) parentFiberRefs
-              else {
-                val newEntry = parentStack match {
-                  case StackEntry(parentFiberId, _, parentVersion) :: tail =>
-                    if (parentFiberId == fiberId)
-                      Value(::(StackEntry(parentFiberId, newValue, parentVersion + 1), tail), parentDepth)
-                    else
-                      Value(::(StackEntry(fiberId, newValue, 0), parentStack), parentDepth + 1)
-                }
-                parentFiberRefs.updated(ref, newEntry)
-              }
+          type V = ref.Value with AnyRef
+          val parentStack = value1.stack
+          val parentDepth = value1.depth
+
+          val ancestor = findAncestor(ref, parentStack, parentDepth, childStack, childDepth)
+          val patch    = ref.diff(ancestor, childValue)
+
+          val oldValue = parentStack.head.value.asInstanceOf[V]
+          val newValue = ref.join(oldValue, ref.patch(patch)(oldValue)).asInstanceOf[V]
+
+          if (oldValue ne newValue) {
+            val newEntry = parentStack match {
+              case StackEntry(parentFiberId, _, parentVersion) :: tail =>
```

Comment:
Not confident as I was unable to decompile the resulting class files, but we may want to avoid the pattern match to prevent the need for two `unapply` invocations.